### PR TITLE
MAINT: resolve UPDATEIFCOPY deprecation errors

### LINKED ISSUE #8145: BUG: ndimage geometric_transform and zoom using deprecated NumPy calls
I'm experiencing an unrelated [Travis failure](https://travis-ci.org/scipy/scipy/jobs/300367861#L1076) over on #8144 related to `ndimage` using deprecated NumPy functionality:
```
        _nd_image.geometric_transform(filtered, None, coordinates, None, None,
>                                     output, order, mode, cval, None, None)
E       DeprecationWarning: NPY_ARRAY_UPDATEIFCOPY, NPY_ARRAY_INOUT_ARRAY, and NPY_ARRAY_INOUT_FARRAY are deprecated, use NPY_WRITEBACKIFCOPY, NPY_ARRAY_INOUT_ARRAY2, or NPY_ARRAY_INOUT_FARRAY2 respectively instead, and call PyArray_ResolveWritebackIfCopy before the array is deallocated, i.e. before the last call to Py_DECREF.
```
And then later:
```
            _nd_image.zoom_shift(filtered, matrix, offset/matrix, output, order,
>                                mode, cval)
E           DeprecationWarning: NPY_ARRAY_UPDATEIFCOPY, NPY_ARRAY_INOUT_ARRAY, and NPY_ARRAY_INOUT_FARRAY are deprecated, use NPY_WRITEBACKIFCOPY, NPY_ARRAY_INOUT_ARRAY2, or NPY_ARRAY_INOUT_FARRAY2 respectively instead, and call PyArray_ResolveWritebackIfCopy before the array is deallocated, i.e. before the last call to Py_DECREF.
```
Any `ndimage` folks have time to look?

### PR DESCRIPTION
**Pull request by ghost**
Closes #8145.

### CODE REVIEW

**Comment by xoviat** [comment]
This should be ready to go. It will start working right after the API version is bumped.

**Comment by xoviat** [comment]
See here for the passing tests: https://travis-ci.org/scipy/scipy/jobs/301100231

File: scipy/ndimage/src/nd_image.c [hunk 1]
```diff
@@ -190,6 +194,9 @@ static PyObject *Py_Correlate1D(PyObject *obj, PyObject *args)
 
     NI_Correlate1D(input, weights, axis, output, (NI_ExtendMode)mode, cval,
                    origin);
+    #if NPY_API_VERSION >= 0x0000000c
+        PyArray_ResolveWritebackIfCopy(output);
+    #endif
```
Comment (tylerjereddy) [inline scipy/ndimage/src/nd_image.c]: This C preprocessor if / endif block is used quite a lot in this PR in different places. Could it perhaps be helpful from a maintenance / code duplication standpoint to abstract the preprocessor code to a single macro declaration at the top of the C module, or perhaps even better in one of the associated header files that is included?

**Review by tylerjereddy — COMMENTED** [review]
My review simply suggests cleaning up the C preprocessor code a little bit.

It may also be helpful to briefly summarize why the current Travis test matrix can't pass with this PR if there's some strange reason for that (I understand there's some tricky low level stuff going on so maybe just a quick note on that so we know it is ok or not).

File: scipy/ndimage/src/nd_image.c (hunk 1 above)
Comment (pv) [reply scipy/ndimage/src/nd_image.c]: Yes, `HAVE_WRITEBACKIFCOPY` please

**Comment by xoviat** [comment]
The Travis-CI is not passing because of numpy/numpy#10014. There is no action required in SciPy.

File: scipy/ndimage/src/nd_image.c [hunk 2]
```diff
@@ -45,6 +45,10 @@
 #include "ccallback.h"
 #include "numpy/npy_3kcompat.h"
 
+#if NPY_API_VERSION >= 0x0000000c
+    #define HAVE_WRITEBACKIFCOPY
+#endif
+
```
Comment (tylerjereddy) [inline scipy/ndimage/src/nd_image.c]: I don't think this really changes much. Rather, I think the idea is to define a macro here. Probably something like this:

```C
#define HAVE_WRITEBACKIFCOPY(output)  if ( NPY_API_VERSION >= 0x0000000c ) \
                                         PyArray_ResolveWritebackIfCopy(output);
```

And then just a single `HAVE_WRITEBACKIFCOPY(output)` statement should suffice at the various locations where there are the `ifdef` blocks I think. Just have to be careful with the preprocessor literal substitutions (my syntax might be slightly off).

Comment (xoviat) [reply scipy/ndimage/src/nd_image.c]: That's not going to work, as the symbol will not be resolved at compile-time.

**Comment by pv** [comment]
I think the ifdefs are ok, better not try to abstract it away too much.

**Comment by pv** [comment]
scipy/sparse/sparsetools/sparsetools.cxx also needs similar change.
The correct place for writeback is here 
https://github.com/scipy/scipy/blob/master/scipy/sparse/sparsetools/sparsetools.cxx#L444 
along the lines of `if (is_output[j] && arg_arrays[j] != NULL) { 
PyArray_WriteBackIfCopy(arg_arrays[j]); }` before the XDECREF.

**Comment by xoviat** [comment]
@pv Done.

**Comment by xoviat** [comment]
Hopefully that will fix the stupid typos that I made.

**Review by tylerjereddy — APPROVED** [review]
Ok, the full CI suite is passing & I see that upstream numpy fixes have been applied as well. I'm +1 to merge as long as @pv is happy with the preprocessor adjustments (looks like his comments were addressed).

**Comment by pv** [comment]
Thanks.

Code style nit: typically `#` preprocessor macros don't follow indentation rules, `#` is first character on a line.

**Comment by xoviat** [comment]
> Code style nit: typically # preprocessor macros don't follow indentation rules, # is first character on a line.

I'm aware of that, and it's very frustrating. When I'm reading C code, I always have to spend an extra few seconds matching the preprocessor definitions. Who came up with this rule?

**Comment by rgommers** [comment]
backporting this to 1.0.x, it fixes a lot of test errors there

**Pull request merged**